File: Stack/962.Maximum-Width-Ramp/962.Maximum-Width-Ramp.py
class Solution:
    def maxWidthRamp(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        Stack = []
        for i in range(len(A)):
            if len(Stack)==0 or A[Stack[-1]]>=A[i]:
                Stack.append(i)
        result = 0
        for i in range(len(A)-1,-1,-1):
            while (len(Stack)>0 and A[Stack[-1]]<=A[i]):
                result = max(result,i-Stack[-1])
                Stack.pop()
        return result
            

File: Stack/962.Maximum-Width-Ramp/962.Maximum-Width-Ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) 
    {
        vector<int>q;
        vector<int>idx;
        int result = 0;
        
        for (int i=0; i<nums.size();i++)
        {           
            if (q.size()==0 || q.size()>0 && nums[i]<q.back())
            {
                q.push_back(nums[i]);
                idx.push_back(i);
            }
            else
            {
                auto it = upper_bound(q.rbegin(),q.rend(),nums[i]);
                int k = it-q.rbegin();
                k = q.size()-1-k + 1;
                result = max(result, i-idx[k]);                
            }            
        }        
        return result;
    }
};
